### PR TITLE
[FLINK-24830][examples] Update DataStream WordCount example

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

Updates the DataStream word count example to better reflect the current state of the art in Apache Flink; in particular the batch stream unification. 

Because of how important this example is, it has more comments than normal to walk users through the application and to show off best practices. 

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by sjwiesman** [comment]
@alibahadirzeybek would you want to take a look at this? The actual code changes are trivial, but I'd be interested in your opinion on the code comments.

**Comment by alpreu** [comment]
@flinkbot run azure

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala [hunk 1]
```diff
@@ -18,74 +18,170 @@
...
+    // The “classic” execution behavior of the DataStream API is called STREAMING execution
+    // mode. Applications should use streaming execution for unbounded jobs that require
+    // continuous incremental processing and are expected to stay online indefinitely.
+    //
+    // By enabling BATCH execution, we allow Flink to apply additional optimizations that we
+    // can only do when we know that our input is bounded. For example, different
+    // join/aggregation strategies can be used, in addition to a different shuffle
+    // implementation that allows more efficient task scheduling and failure recovery behavior.
+    //
+    // By setting the runtime mode to AUTOMATIC, Flink will choose the BATCH  if all sources
```
Comment (alpreu) [inline flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:91]:
```suggestion
    // By setting the runtime mode to AUTOMATIC, Flink will choose BATCH if all sources
```

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala [hunk 2]
```diff
@@ -18,74 +18,170 @@
...
+    // are bounded and otherwise STREAMING.
```
Comment (alpreu) [inline flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:110]: Duplicate comment

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/WordCount.java [hunk 3]
```diff
@@ -50,51 +69,92 @@
...
+        // The “classic” execution behavior of the DataStream API is called STREAMING execution
+        // mode. Applications should use streaming execution for unbounded jobs that require
+        // continuous incremental processing and are expected to stay online indefinitely.
+        //
+        // By enabling BATCH execution, we allow Flink to apply additional optimizations that we
+        // can only do when we know that our input is bounded. For example, different
+        // join/aggregation strategies can be used, in addition to a different shuffle
+        // implementation that allows more efficient task scheduling and failure recovery behavior.
+        //
+        // By setting the runtime mode to AUTOMATIC, Flink will choose the BATCH  if all sources
```
Comment (alpreu) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/WordCount.java:94]:
```suggestion
        // By setting the runtime mode to AUTOMATIC, Flink will choose BATCH if all sources
```

**Review by alpreu — CHANGES_REQUESTED** [review]
Thank you for your contribution. I checked the code and it looks good to me, I only found some small issues in the comments

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala (hunk 2 above)
Comment (sjwiesman) [reply flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:110]: whoops 😅

**Comment by sjwiesman** [comment]
@alpreu thanks for the quick review! I'll apply your fixes while merging :)

**Comment by sjwiesman** [comment]
It looks like there are some relevant test failures because some tests actually use the examples as test dependencies. I'll look into it.

**Comment by sjwiesman** [comment]
Nevermind, it's because my changes to the test base weren't filtering out hidden files. I'll merge if green.

**Comment by sjwiesman** [comment]
@gaborgsomogyi or @KarmaGYZ could either of you please take a look at the failing Yarn tests and help me debug them? I have made a simple change to the DataStream WordCount example, which is used by the Yarn tests as the submitted application. 

After speaking offline with @AHeise and @tillrohrmann we believe this may be due to the added dependency on `flink-connector-files` which is causing some sort of version clash with Yarn. Any insights on how to resolve this would be greatly appreciated.

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala [hunk 4]
```diff
@@ -18,74 +18,154 @@
...
-    // emit result
-    if (params.has("output")) {
-      counts.writeAsText(params.get("output"))
-    } else {
-      println("Printing result to stdout. Use --output to specify output path.")
-      counts.print()
+    val counts =
+      // The text lines read from the source are split into words
+      // using a user-defined function. The tokenizer, implemented below,
+      // will output each words as a (2-tuple) containing (word, 1)
```
Comment (alibahadirzeybek) [inline flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:120]:
```suggestion
      // will output each word as a (2-tuple) containing (word, 1)
```

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/util/CLI.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    public static CLI fromArgs(String[] args) throws Exception {
+        MultipleParameterTool params = MultipleParameterTool.fromArgs(args);
+        Path[] inputs = null;
+        if (params.has(INPUT_KEY)) {
+            inputs =
+                    params.getMultiParameterRequired(INPUT_KEY).stream()
+                            .map(Path::new)
+                            .toArray(Path[]::new);
+        } else {
+            System.out.println("Executing example with default input data set.");
```
Comment (alibahadirzeybek) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/util/CLI.java:54]:
```suggestion
            System.out.println("Executing example with default input data.");
```

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala [hunk 6]
```diff
@@ -18,74 +18,154 @@
...
+      text.flatMap(new Tokenizer)
+        .name("tokenizer")
+        // keyBy groups tuples based on the "_1" field, the word.
+        // Using a keyBy allows performing aggregations and other
+        // stateful transformations over data on a per-key basis.
+        // This is similar to a GROUP BY clause in a SQL query.
+        .keyBy(_._1)
+        // For each key, we perform a simple sum of the "1" field, the count.
+        // If the input data set is bounded, sum will output a final count for
```
Comment (alibahadirzeybek) [inline flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:129]:
```suggestion
        // If the input data stream is bounded, sum will output a final count for
```

File: flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala [hunk 7]
```diff
@@ -18,74 +18,154 @@
...
+    // submits the Job and begins processing.
+    env.execute("WordCount")
+  }
+
+  // *************************************************************************
+  // USER FUNCTIONS
+  // *************************************************************************
+
+  /**
+   * Implements the string tokenizer that splits sentences into words as a user-defined
```
Comment (alibahadirzeybek) [inline flink-examples/flink-examples-streaming/src/main/scala/org/apache/flink/streaming/scala/examples/wordcount/WordCount.scala:161]:
```suggestion
   * Implements the string tokenizer that splits a sentence into words as a user-defined
```

Same for the Java implementation

File: flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/WordCount.java [hunk 8]
```diff
@@ -50,51 +70,92 @@
...
-                        // group by the tuple field "0" and sum up tuple field "1"
+                        .name("tokenizer")
+                        // keyBy groups tuples based on the "0" field, the word.
+                        // Using a keyBy allows performing aggregations and other
+                        // stateful transformations over data on a per-key basis.
+                        // This is similar to a GROUP BY clause in a SQL query.
                         .keyBy(value -> value.f0)
-                        .sum(1);
+                        // For each key, we perform a simple sum of the "1" field, the count.
+                        // If the input data set is bounded, sum will output a final count for
```
Comment (alibahadirzeybek) [inline flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/wordcount/WordCount.java:132]:
```suggestion
                        // If the input data stream is bounded, sum will output a final count for
```

**Review by alibahadirzeybek — COMMENTED** [review]
Suggested a couple of changes to wordings in some of the comments, otherwise LGTM.

File: flink-examples/flink-examples-streaming/pom.xml [hunk 9]
```diff
@@ -147,16 +152,6 @@ under the License.
...
-								<!-- For WordCount example data -->
-								<artifactItem>
-									<groupId>org.apache.flink</groupId>
-									<artifactId>flink-examples-batch_${scala.binary.version}</artifactId>
-									<version>${project.version}</version>
-									<type>jar</type>
-									<overWrite>false</overWrite>
-									<outputDirectory>${project.build.directory}/classes</outputDirectory>
-									<includes>org/apache/flink/streaming/examples/wordcount/util/WordCountData.class</includes>
-								</artifactItem>
```
Comment (sjwiesman) [inline flink-examples/flink-examples-streaming/pom.xml:159]: Turns out that copying this class overwrites the `wordcount/util` folder and deletes CLI.class

File: flink-examples/flink-examples-streaming/pom.xml [hunk 10]
```diff
@@ -261,32 +256,6 @@ under the License.
...
-							<includes>
-								<include>org/apache/flink/streaming/examples/wordcount/PojoExample.class</include>
-								<include>org/apache/flink/streaming/examples/wordcount/PojoExample$*.class</include>
-								<include>org/apache/flink/streaming/examples/wordcount/util/WordCountData.class</include>
-								<include>META-INF/LICENSE</include>
-								<include>META-INF/NOTICE</include>
-							</includes>
-						</configuration>
-					</execution>
-
```
Comment (sjwiesman) [inline flink-examples/flink-examples-streaming/pom.xml:289]: This job doesn't exist anymore

**Comment by MartijnVisser** [comment]
@sjwiesman Should we also update the example in the README?

**Comment by sjwiesman** [comment]
@MartijnVisser I have an umbrella ticket, FLINK-24635, for cleaning up the examples. You can open that as another subtask though I think the entire README could use a rewrite. It still shows off DataSet and does not have a table API example.

**Pull request closed without merging**